iceberg/src/main/scala/com/nvidia/spark/rapids/iceberg/parquet/reader.scala

```diff
@@ -0,0 +1,248 @@
+            dictFilter.shouldRead(typeWithIds, rowGroup, reader.getDictionaryReader(rowGroup)) &&
+            bloomFilter.shouldRead(typeWithIds, rowGroup, reader.getBloomFilterDataReader(rowGroup))
+      }
+    }.getOrElse(blocks)
+  }
+
+  def clipBlocksToSchema(fileReadSchema: ShadedMessageType,
+      blocks: Seq[ShadedBlockMetaData]): Seq[BlockMetaData] = {
+    GpuParquetUtils.clipBlocksToSchema(unshade(fileReadSchema),
+      blocks.map(unshade).asJava,
+      conf.caseSensitive)
+  }
+
+  def filterParquetBlocks(file: IcebergPartitionedFile,
+      requiredSchema: Schema): ParquetFileInfoWithBlockMeta = {
+    withResource(file.newReader) { reader =>
+      val fileSchema = reader.getFileMetaData.getSchema
+
+      val rowGroupFirstRowIndices = new Array[Long](reader.getRowGroups.size())
+      rowGroupFirstRowIndices(0) = 0
+      var accumulatedRowCount = reader.getRowGroups.get(0).getRowCount
+      for (i <- 1 until reader.getRowGroups.size()) {
+        rowGroupFirstRowIndices(i) = accumulatedRowCount
+        accumulatedRowCount += reader.getRowGroups.get(i).getRowCount
+      }
+      val (typeWithIds, fileReadSchema) = projectSchema(fileSchema, requiredSchema)
+      val filteredBlocks = filterRowGroups(reader, requiredSchema, typeWithIds, file.filter)
+      val blockFirstRowIndices = filteredBlocks.map(b => rowGroupFirstRowIndices(b.getOrdinal))
+      val blocks = clipBlocksToSchema(fileReadSchema, filteredBlocks)
+
+      val partReaderSparkSchema = TypeWithSchemaVisitor.visit(requiredSchema.asStruct(),
+          fileReadSchema, new SparkSchemaConverter)
+        .asInstanceOf[StructType]
+
+      ParquetFileInfoWithBlockMeta(file.path,
+        blocks,
+        InternalRow.empty, // Iceberg handles partition values but itself
+        unshade(fileReadSchema),
+        partReaderSparkSchema,
+        DateTimeRebaseCorrected,
+        DateTimeRebaseCorrected,
+        hasInt96Timestamps = true,
+        blockFirstRowIndices,
+      )
+    }
+  }
+}
+
+object GpuIcebergParquetReader {
+  private val READ_PROPERTIES_TO_REMOVE = Set(
+    "parquet.read.filter",
+    "parquet.private.read.filter.predicate",
+    "parquet.read.support.class")
+
+  def buildReaderOptions(file: InputFile, split: Option[(Long, Long)])
+  : ParquetReadOptions = {
+    var optionsBuilder: ParquetReadOptions.Builder = null
+    file match {
+      case hadoop: HadoopInputFile =>
+        // remove read properties already set that may conflict with this read
+        val conf = new Configuration(hadoop.getConf)
+        for (property <- READ_PROPERTIES_TO_REMOVE) {
+          conf.unset(property)
+        }
+        optionsBuilder = HadoopReadOptions.builder(conf)
+      case _ =>
+        throw new UnsupportedOperationException("Only Hadoop files are supported for now")
+    }
+    split.foreach { case (start, length) =>
+      optionsBuilder = optionsBuilder.withRange(start, start + length)
+    }
+    optionsBuilder.build
```

Comment:
If split is empty, then will trigger NPE becasue `optionsBuilder` is null. Is this expected?